apps/HelloAndroidCamera2/src/com/example/helloandroidcamera2/Camera2BasicFragment.java

```diff
@@ -0,0 +1,568 @@
+import android.content.DialogInterface;
+import android.graphics.ImageFormat;
+import android.hardware.camera2.CameraAccessException;
+import android.hardware.camera2.CameraCaptureSession;
+import android.hardware.camera2.CameraCharacteristics;
+import android.hardware.camera2.CameraDevice;
+import android.hardware.camera2.CameraManager;
+import android.hardware.camera2.CaptureRequest;
+import android.hardware.camera2.CaptureResult;
+import android.hardware.camera2.TotalCaptureResult;
+import android.hardware.camera2.params.StreamConfigurationMap;
+import android.media.Image;
+import android.media.ImageReader;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.Message;
+import android.util.Log;
+import android.util.Size;
+import android.view.LayoutInflater;
+import android.view.Surface;
+import android.view.SurfaceHolder;
+import android.view.SurfaceView;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Toast;
+
+import com.android.ex.camera2.blocking.BlockingCameraManager;
+import com.android.ex.camera2.blocking.BlockingCameraManager.BlockingOpenException;
+import com.android.ex.camera2.blocking.BlockingSessionCallback;
+import com.android.ex.camera2.blocking.BlockingStateCallback;
+import com.android.ex.camera2.exceptions.TimeoutRuntimeException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Semaphore;
+
+public class Camera2BasicFragment extends Fragment implements View.OnClickListener {
+
+    /**
+     * Tag for the {@link Log}.
+     */
+    private static final String TAG = "Camera2BasicFragment";
+
+    private static final int STATE_TIMEOUT_MS = 5000;
+    private static final int SESSION_WAIT_TIMEOUT_MS = 2500;
+
+    private static final Size IMAGE_READER_SIZE = new Size(1920, 1440);
+    private static final int IMAGE_READER_BUFFER_SIZE = 16;
+
+    private final SurfaceHolder.Callback mSurfaceCallback
+            = new SurfaceHolder.Callback() {
+
+        @Override
+        public void surfaceCreated(SurfaceHolder holder) {
+            // The surface is ready. Make its buffers use the YV12 format and use the NDK to
+            // reconfigure its buffers to be of a different size than on screen (i.e., use the
+            // resolution of the ImageReader instead and use the hardware scaler to interpolate).
+            // (If you don't set it here, funny things happen when if you sleep the device).
+            holder.setFormat(ImageFormat.YV12);
+            JNIUtils.configureSurface(holder.getSurface(),
+                    mImageReader.getWidth(), mImageReader.getHeight());
+            mSurface = holder.getSurface();
+            openCamera();
+        }
+
+        @Override
+        public void surfaceChanged(SurfaceHolder holder, int format, int width,
+                int height) {
+
```

Comment:
If these are deliberately empty, it's probably good form to add a one-line comment to that effect